R/modify-records.R

```diff
@@ -0,0 +1,151 @@
+#' Modify options and records from a `NONMEM` control stream file
+#' @param .mod a bbr model object
+#' @name modify_records
+#'
+#' @details
+#'  - **`safe_read_ctl()`** is called internally within the other functions, though
+#'  it can also be used outside of that context.
+#'  - **`modify_prob_statement()`**, **`modify_data_path_ctl()`**, and
+#'   **`remove_records()`** read in the control stream, make any modifications,
+#'   and then save out the updated control stream.
+#'     - `modify_prob_statement()` also returns a character string defining the
+#'     `$PROBLEM` text (see `prob_text` argument).
+#' - **`mod_has_record()`** will return a logical value denoting whether a `bbr`
+#'   model has a given record type.
+NULL
+
+#' @describeIn modify_records Safely read in a `NONMEM` control stream file via
+#' `nmrec`
+#' @keywords internal
+safe_read_ctl <- function(.mod){
+  check_model_object(.mod, c(NM_MOD_CLASS, NMBOOT_MOD_CLASS))
+  mod_path <- get_model_path(.mod)
+  ctl <- tryCatch(nmrec::read_ctl(mod_path), error = identity)
+  if(inherits(ctl, "error")){
+    warning(
+      glue::glue("Could not read control stream file. Reason: {ctl$message}")
+    )
+  }
+  return(ctl)
+}
+
+
+#' @describeIn modify_records Modify or retrieve the `$PROBLEM` statement from
+#' a `NONMEM` control stream file.
+#'
+#' @param prob_text If `NULL` return the current `$PROBLEM` statement. If a
+#'  character string, set the problem statement to that value.
+#' @keywords internal
+modify_prob_statement <- function(.mod, prob_text = NULL){
+  mod_path <- get_model_path(.mod)
+  ctl <- safe_read_ctl(.mod)
+
+  prob_recs <- nmrec::select_records(ctl, "prob")
+  if (length(prob_recs) != 1) {
+    dev_error("select_records should always return a single $PROBLEM record")
+  }
+
+  prob_rec <- prob_recs[[1]]
+  prob_str <- nmrec::get_record_option(prob_rec, "text")
+  if(is.null(prob_str)){
+    # NONMEM permits a bare "$PROB{newline}", in which case there won't be a text option.
+    prob_rec$parse()
+    idx <- purrr::detect_index(
+      prob_rec$values,
+      function(x) inherits(x, "nmrec_option_record_name")
+    )
+    if(idx == 0){
+      dev_error("Unsupported $PROBLEM record format")
+    }
+    if(!is.null(prob_text)){
+      prob_rec$values <- append(prob_rec$values, paste0(" ", prob_text), idx)
+      prob_str_return <- prob_rec$values[[idx + 1]]$format()
+    }else{
+      prob_str_return <- prob_rec$values[[idx]]$format()
```

Comment:
Could have done it a different way (rather than `NULL`), but I was ultimately trying to reduce the number of times we do certain checks (such as a newline problem statement or checking for multiple statements), and therefore opted to contain the reading and modification within the same function. When it comes to the purpose of the reading: the intention was to pull each of these statements out for storing in the bootstrap spec, including in a `run_log`, etc.